pkg/nodeagent/registry/containerd_extractor.go

```diff
@@ -0,0 +1,165 @@
+func NewExtractor() Extractor {
+	return &containerdExtractor{}
+}
+
+// CopyFromImage copies files from a given image reference to the destination folder.
+func (e *containerdExtractor) CopyFromImage(ctx context.Context, imageRef string, files []string, destination string) error {
+	address := os.Getenv("CONTAINERD_ADDRESS")
+	if address == "" {
+		address = defaults.DefaultAddress
+	}
+	namespace := os.Getenv(namespaces.NamespaceEnvVar)
+	if namespace == "" {
+		namespace = namespaces.Default
+	}
+
+	client, err := containerd.New(address, containerd.WithDefaultNamespace(namespace))
+	if err != nil {
+		return fmt.Errorf("error creating containerd client: %w", err)
+	}
+	ctx, done, err := client.WithLease(ctx)
+	if err != nil {
+		return fmt.Errorf("error adding lease to containerd client: %w", err)
+	}
+	defer func() { utilruntime.HandleError(done(ctx)) }()
+
+	image, err := client.Pull(ctx, imageRef, containerd.WithPullSnapshotter(containerd.DefaultSnapshotter), containerd.WithPullUnpack)
+	if err != nil {
+		return fmt.Errorf("error pulling image: %w", err)
+	}
+
+	snapshotter := client.SnapshotService(containerd.DefaultSnapshotter)
+
+	imageMountDirectory, err := os.MkdirTemp("", "node-agent-")
+	if err != nil {
+		return fmt.Errorf("error creating temp image: %w", err)
+	}
+	defer os.Remove(imageMountDirectory)
+
+	if err := mountImage(ctx, image, snapshotter, imageMountDirectory); err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		sourceFile := path.Join(imageMountDirectory, file)
+		destinationFile := path.Join(destination, path.Base(file))
+		if err := CopyFile(sourceFile, destinationFile); err != nil {
+			return fmt.Errorf("error copying file %s to %s: %w", sourceFile, destinationFile, err)
+		}
+	}
+
+	return unmountImage(ctx, snapshotter, imageMountDirectory)
+}
+
+func mountImage(ctx context.Context, image containerd.Image, snapshotter snapshots.Snapshotter, directory string) error {
+	diffIDs, err := image.RootFS(ctx)
+	if err != nil {
+		return err
+	}
+	chainID := identity.ChainID(diffIDs).String()
+
+	mounts, err := snapshotter.View(ctx, directory, chainID)
+	if err != nil {
+		return err
+	}
+
+	if err := mount.All(mounts, directory); err != nil {
+		if err := snapshotter.Remove(ctx, directory); err != nil && !errdefs.IsNotFound(err) {
+			return fmt.Errorf("error cleaning up snapshot after mount error: %w", err)
+		}
+		return fmt.Errorf("error mounting image: %w", err)
+	}
+
+	return nil
+}
+
+func unmountImage(ctx context.Context, snapshotter snapshots.Snapshotter, directory string) error {
+	if err := mount.UnmountAll(directory, 0); err != nil {
+		return err
+	}
+
+	if err := snapshotter.Remove(ctx, directory); err != nil && !errdefs.IsNotFound(err) {
+		return fmt.Errorf("error removing snapshot: %w", err)
+	}
+
+	return nil
+}
+
+// CopyFile copies a source file to destination file and sets the executable flag.
+func CopyFile(sourceFile, destinationFile string) error {
```

Comment:
Can you use an `afero.Fs` here?

Reply:
Isn't this too complex? For `afero.Fs` the function needs to be converted into a method that you can exchange the FS for tests etc.
From my point of view `afero.Fs` is valuable for tests of a controller with file system operations, which could have side effects to the hosts file system. Here, it is just copying from A to B which can be easily tested using temporary files and folders.

Reply:
Why complex - don't you just need to provide the `afero.Fs` as parameter to the function and then call the FS operations from there instead of using the `os` package?

I'm suggesting this because we might need it later when we integrate this code into the `operatingsystemconfig` controller - here, we already have an `afero.Fs` and wouldn't need to use a second mechanism (your fake extractor also calls this function and would manipulate the host OS then).